rust/xaynet-client/src/api/http.rs

```diff
@@ -21,13 +23,65 @@ pub struct HttpApiClient {
 }
 
 impl HttpApiClient {
-    pub fn new<S>(address: S) -> Self
+    /// Creates a new HTTP(S) client.
+    ///
+    /// # Warning
+    /// If no trusted root `certificate`s are provided, then every server will be trusted.
+    pub fn new<S>(
+        address: S,
+        certificates: Option<Vec<Certificate>>,
+    ) -> Result<Self, HttpApiClientError>
     where
         S: Into<String>,
     {
-        Self {
-            client: Client::new(),
+        let client = if let Some(certificates) = certificates {
+            let mut builder = ClientBuilder::new().use_rustls_tls();
+            for certificate in certificates {
+                builder = builder.add_root_certificate(certificate);
+            }
+            builder.build().map_err(HttpApiClientError::Http)?
+        } else {
+            ClientBuilder::new()
+                .use_rustls_tls()
+                .danger_accept_invalid_certs(true)
+                .build()
+                .map_err(HttpApiClientError::Http)?
+        };
+
+        Ok(Self {
+            client,
             address: address.into(),
+        })
+    }
+
+    /// Reads DER and PEM certificates from given paths.
+    pub fn certificates_from(
+        paths: &Option<Vec<PathBuf>>,
+    ) -> Result<Option<Vec<Certificate>>, HttpApiClientError> {
+        if let Some(paths) = paths {
+            if paths.is_empty() {
+                Ok(None)
+            } else {
+                paths
+                    .iter()
+                    .map(|path| -> Result<Certificate, HttpApiClientError> {
+                        let encoding = fs::read(path).map_err(HttpApiClientError::Io)?;
+                        path.extension().map_or_else(
+                            || Err(HttpApiClientError::UnexpectedCertificate),
+                            |extension| match extension.to_str() {
+                                Some("der") => Certificate::from_der(&encoding)
+                                    .map_err(HttpApiClientError::Http),
+                                Some("pem") => Certificate::from_pem(&encoding)
+                                    .map_err(HttpApiClientError::Http),
+                                _ => Err(HttpApiClientError::UnexpectedCertificate),
+                            },
+                        )
+                    })
+                    .collect::<Result<Vec<_>, HttpApiClientError>>()
```

Comment:
Nice I didn't know `collect` could do that. I assume it stops on the first error?

Reply:
it short-circuits and returns the first encountered error, that's pretty neat indeed :) should work similarly with vecs and options as well.

Reply:
`Iterator` is a little gem ^^